sharktank/tests/models/llama/sharded_llama_test.py

```diff
@@ -5,122 +5,119 @@
             int(torch.max(seq_lens)), model.cache.pad_sequence_stride
         )
         token_ids = torch.randint(
             low=0,
-            high=vocabulary_size,
-            size=[batch_size, batch_seq_len],
+            high=self.vocabulary_size,
+            size=[self.batch_size, batch_seq_len],
             dtype=torch.int32,
         )
         attention_mask = model.attention_mask(model.input_mask(seq_lens, batch_seq_len))
         seq_block_ids = torch.arange(
-            batch_size * batch_seq_len // config.block_seq_stride
-        ).view(batch_size, -1)
+            self.batch_size * batch_seq_len // self.config.block_seq_stride
+        ).view(self.batch_size, -1)
+        cache_state = model.cache.paged.allocate(page_count=self.cache_page_count)
+        cache_state = [torch.rand_like(cache_state[0])]
+        return {
+            "tokens": token_ids,
+            "attention_mask": attention_mask,
+            "seq_block_ids": seq_block_ids,
+            "cache_state": cache_state,
+        }
 
-        # Verify prefill step.
-        sharded_config = deepcopy(config)
-        sharded_config.tensor_parallelism_size = 2
-        sharded_theta = shard_theta(theta, sharded_config)
-        sharded_model = PagedLlamaModelV1(sharded_theta, sharded_config)
+    def make_equal_unsharded_and_sharded_prefill_args(
+        self, model: PagedLlamaModelV1, sharded_model: PagedLlamaModelV1
+    ) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+        prefill_args = self.make_prefill_args(model)
         sharded_cache_state = sharded_model.cache.paged.allocate(
-            page_count=cache_page_count
+            page_count=self.cache_page_count
+        )
+        assert iterables_equal(
+            prefill_args["cache_state"][0].shape, sharded_cache_state[0].shape
         )
+        sharded_prefill_args = deepcopy(prefill_args)
         sharded_cache_state = sharded_model.cache.paged.shard_state(
-            deepcopy(cache_state)
+            sharded_prefill_args["cache_state"]
         )
+        sharded_prefill_args["cache_state"] = sharded_cache_state
+        return prefill_args, sharded_prefill_args
 
-        expected_prefill_result = model.prefill(
-            token_ids,
-            attention_mask=attention_mask,
-            seq_block_ids=seq_block_ids,
-            cache_state=cache_state,
-        )
-        sharded_prefill_result = sharded_model.prefill(
-            token_ids,
-            attention_mask=attention_mask,
-            seq_block_ids=seq_block_ids,
-            cache_state=sharded_cache_state,
-        )
-        # The errors are quite high, but for float64 both errors drop to < 1e-12.
-        # The numerics are probably correct.
-        torch.testing.assert_close(
-            sharded_prefill_result, expected_prefill_result, atol=1e-3, rtol=1e-2
+    def make_decode_args(self, model: PagedLlamaModelV1) -> Dict[str, Any]:
+        seq_lens = torch.randint(
```

Comment:
Be specific in random testing where possible. If we want to check variable sequence length we should pick a set of numbers instead of randomizing.

Reply:
We are setting the random seed at the test setup stage, so it should be consistent between runs. Isn't the torch random generator going to be stable between releases?

Reply:
I changed them to fixed values.